File: app/src/androidTest/java/com/orgzly/android/reminders/ReminderServiceTest.java
package com.orgzly.android.reminders;

import com.orgzly.android.OrgzlyTest;
import com.orgzly.android.db.dao.ReminderTimeDao;
import com.orgzly.android.prefs.AppPreferences;

import org.joda.time.Instant;
import org.joda.time.LocalDateTime;
import org.junit.Test;

import java.util.List;

import static org.junit.Assert.assertEquals;

public class ReminderServiceTest extends OrgzlyTest {
    @Test
    public void testNotesWithTimesSkipDoneState() {
        testUtils.setupBook("notebook",
                "* Note 1\n"+
                "SCHEDULED: <2017-03-20>\n" +
                "* DONE Note 2\n"+
                "SCHEDULED: <2017-03-20>\n" +
                "* Note 3");

        LastRun lastRun = new LastRun();
        Instant now = Instant.parse("2017-03-15");
        AppPreferences.remindersForScheduledEnabled(context, true);

        List<NoteReminder> notes = ReminderService.getNoteReminders(
                context, dataRepository, now, lastRun, ReminderService.TIME_FROM_NOW);

        assertEquals(1, notes.size());
    }

    @Test
    public void testNotesWithTimesWithRepeater() {
        testUtils.setupBook("notebook",
                "* Note 1\n"+
                "SCHEDULED: <2017-03-10 Fri +1w>\n" +
                "* Note 2\n"+
                "SCHEDULED: <2017-03-20 Mon 16:00>\n" +
                "* Note 3\n" +
                "* Note 4\n"+
                "SCHEDULED: <2017-03-16 Fri +1w>\n");

        LastRun lastRun = new LastRun();
        Instant now = Instant.parse("2017-03-15T13:00:00"); // Wed
        AppPreferences.remindersForScheduledEnabled(context, true);

        List<NoteReminder> notes = ReminderService.getNoteReminders(
                context, dataRepository, now, lastRun, ReminderService.TIME_FROM_NOW);

        assertEquals(2, notes.size());

        assertEquals("Note 4", notes.get(0).getPayload().title);
        assertEquals("2017-03-16T09:00:00", new LocalDateTime(notes.get(0).getRunTime()).toString("yyyy-MM-dd'T'HH:mm:ss"));

        assertEquals("Note 2", notes.get(1).getPayload().title);
        assertEquals("2017-03-20T16:00:00", new LocalDateTime(notes.get(1).getRunTime()).toString("yyyy-MM-dd'T'HH:mm:ss"));
    }

    @Test
    public void testReminderForDeadlineTime() {
        testUtils.setupBook("notebook",
                "* Note 1\n"+
                "SCHEDULED: <2017-03-16 Fri +1w>\n" +
                "* Note 2\n"+
                "DEADLINE: <2017-03-20 Mon 16:00>\n");

        LastRun lastRun = new LastRun();
        Instant now = Instant.parse("2017-03-15T13:00:00"); // Wed
        AppPreferences.remindersForDeadlineEnabled(context, true);

        List<NoteReminder> notes = ReminderService.getNoteReminders(
                context, dataRepository, now, lastRun, ReminderService.TIME_FROM_NOW);

        assertEquals(1, notes.size());

        NoteReminder reminder = notes.get(0);
        assertEquals("Note 2", reminder.getPayload().title);
        assertEquals(ReminderTimeDao.DEADLINE_TIME, reminder.getPayload().timeType);
        assertEquals("2017-03-20T16:00:00", new LocalDateTime(reminder.getRunTime()).toString("yyyy-MM-dd'T'HH:mm:ss"));
    }
}

File: app/src/androidTest/java/com/orgzly/android/reminders/ReminderServiceTest.kt
package com.orgzly.android.reminders

import com.orgzly.android.OrgzlyTest
import com.orgzly.android.db.dao.ReminderTimeDao
import com.orgzly.android.reminders.ReminderService.Companion.getNoteReminders
import org.joda.time.Instant
import org.junit.Assert
import org.junit.Test

class ReminderServiceTest : OrgzlyTest() {
    @Test
    fun testNotesWithTimesSkipDoneState() {
        testUtils.setupBook(
                "notebook",
                """
                    * Note 1
                    SCHEDULED: <2017-03-20>
                    * DONE Note 2
                    SCHEDULED: <2017-03-20>
                    * Note 3
                """.trimIndent())

        val now = Instant.parse("2017-03-15")

        val notes = getNoteReminders(
                context, dataRepository, now, LastRun(), ReminderService.TIME_FROM_NOW)

        Assert.assertEquals(1, notes.size.toLong())
    }

    @Test
    fun testNotesWithTimesWithRepeater() {
        testUtils.setupBook(
                "notebook",
                """
                    * Note 1
                    SCHEDULED: <2017-03-10 Fri +1w>
                    * Note 2
                    SCHEDULED: <2017-03-20 Mon 16:00>
                    * Note 3
                    * Note 4
                    SCHEDULED: <2017-03-16 Thu +1w>
                """.trimIndent())

        val now = Instant.parse("2017-03-15T13:00:00") // Wed

        val notes = getNoteReminders(
                context, dataRepository, now, LastRun(), ReminderService.TIME_FROM_NOW)

        Assert.assertEquals(2, notes.size.toLong())

        notes[0].apply {
            Assert.assertEquals("Note 4", payload.title)
            Assert.assertEquals("2017-03-16T09:00:00.000", runTime.toLocalDateTime().toString())
        }

        notes[1].apply {
            Assert.assertEquals("Note 2", payload.title)
            Assert.assertEquals("2017-03-20T16:00:00.000", runTime.toLocalDateTime().toString())
        }
    }

    @Test
    fun testReminderForDeadlineTime() {
        testUtils.setupBook(
                "notebook",
                """
                    * Note 1
                    SCHEDULED: <2017-03-16 Thu +1w>
                    * Note 2
                    DEADLINE: <2017-03-20 Mon 16:00>
                """.trimIndent())

        val now = Instant.parse("2017-03-15T13:00:00") // Wed

        val notes = getNoteReminders(
                context, dataRepository, now, LastRun(), ReminderService.TIME_FROM_NOW)

        Assert.assertEquals(2, notes.size.toLong())

        notes[0].apply {
            Assert.assertEquals("Note 1", payload.title)
            Assert.assertEquals(ReminderTimeDao.SCHEDULED_TIME, payload.timeType)
            Assert.assertEquals("2017-03-16T09:00:00.000", runTime.toLocalDateTime().toString())
        }

        notes[1].apply {
            Assert.assertEquals("Note 2", payload.title)
            Assert.assertEquals(ReminderTimeDao.DEADLINE_TIME, payload.timeType)
            Assert.assertEquals("2017-03-20T16:00:00.000", runTime.toLocalDateTime().toString())
        }
    }

    @Test
    fun testSchedulingBeforeDailyTime() {
        testUtils.setupBook(
                "notebook",
                """
                    * Note 1
                    SCHEDULED: <2017-03-16 Fri +1w>
                """.trimIndent())

        val now = Instant.parse("2017-03-16T07:00:00")

        val notes = getNoteReminders(
                context, dataRepository, now, LastRun(), ReminderService.TIME_FROM_NOW)

        Assert.assertEquals(1, notes.size.toLong())

        notes[0].apply {
            Assert.assertEquals("Note 1", payload.title)
            Assert.assertEquals(ReminderTimeDao.SCHEDULED_TIME, payload.timeType)
            Assert.assertEquals("2017-03-16T09:00:00.000", runTime.toLocalDateTime().toString())
        }
    }
}